lib/Conversion/CGGIToJaxite/CGGIToJaxite.cpp

```diff
@@ -0,0 +1,281 @@
+#include "lib/Dialect/CGGI/IR/CGGIOps.h"
+#include "lib/Dialect/Jaxite/IR/JaxiteDialect.h"
+#include "lib/Dialect/Jaxite/IR/JaxiteOps.h"
+#include "lib/Dialect/Jaxite/IR/JaxiteTypes.h"
+#include "lib/Dialect/LWE/IR/LWEDialect.h"
+#include "lib/Dialect/LWE/IR/LWEOps.h"
+#include "lib/Dialect/LWE/IR/LWETypes.h"
+#include "llvm/include/llvm/ADT/SmallVector.h"           // from @llvm-project
+#include "llvm/include/llvm/Support/Casting.h"           // from @llvm-project
+#include "llvm/include/llvm/Support/ErrorHandling.h"     // from @llvm-project
+#include "mlir/include/mlir/Dialect/Arith/IR/Arith.h"    // from @llvm-project
+#include "mlir/include/mlir/Dialect/Func/IR/FuncOps.h"   // from @llvm-project
+#include "mlir/include/mlir/Dialect/Tensor/IR/Tensor.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/BuiltinTypeInterfaces.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/BuiltinTypes.h"           // from @llvm-project
+#include "mlir/include/mlir/IR/ImplicitLocOpBuilder.h"   // from @llvm-project
+#include "mlir/include/mlir/IR/PatternMatch.h"           // from @llvm-project
+#include "mlir/include/mlir/IR/Types.h"                  // from @llvm-project
+#include "mlir/include/mlir/IR/Value.h"                  // from @llvm-project
+#include "mlir/include/mlir/IR/Visitors.h"               // from @llvm-project
+#include "mlir/include/mlir/Support/LLVM.h"              // from @llvm-project
+#include "mlir/include/mlir/Support/LogicalResult.h"     // from @llvm-project
+#include "mlir/include/mlir/Transforms/DialectConversion.h"  // from @llvm-project
+
+namespace mlir::heir {
+
+#define GEN_PASS_DEF_CGGITOJAXITE
+#include "lib/Conversion/CGGIToJaxite/CGGIToJaxite.h.inc"
+
+class CGGIToJaxiteTypeConverter : public TypeConverter {
+ public:
+  CGGIToJaxiteTypeConverter(MLIRContext *ctx) {
+    addConversion([](Type type) { return type; });
+    addConversion([ctx](lwe::LWECiphertextType type) -> Type {
+      if (widthFromEncodingAttr(type.getEncoding()) == 3) {
+        return type;
+      }
+      llvm_unreachable("Unsupported cleartext bitwidth in jaxite");
+      return 0;
+    });
+    addConversion([this](ShapedType type) -> Type {
+      return type.cloneWith(type.getShape(),
+                            this->convertType(type.getElementType()));
+    });
+  }
+};
+
+/// Returns true if the func's body contains any CGGI ops.
+bool containsCGGIToJaxiteOps(func::FuncOp func) {
+  auto walkResult = func.walk([&](Operation *op) {
+    if (llvm::isa<cggi::CGGIDialect>(op->getDialect()) ||
+        llvm::isa<lwe::LWEDialect>(op->getDialect()))
+      return WalkResult::interrupt();
+    return WalkResult::advance();
+  });
+  return walkResult.wasInterrupted();
+}
+
+// Returns the Value corresponding to a server key in the FuncOp containing
+// this op.
+template <typename ArgType>
+bool HasArg(func::FuncOp op) {
+  for (auto functionArg : op.getFunctionType().getInputs()) {
+    if (mlir::isa<Type>(functionArg)) {
```

Comment:
Did you mean the template param `<ArgType>` here? It would probably always return true

Reply:
Thanks for catching that. Thought I removed this function in favor of the one-liner from the other comment.